keyboards/bastardkb/charybdis/charybdis.c

```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright 2020 Christopher Courtney <drashna@live.com> (@drashna)
+ * Copyright 2021 Quentin LEBASTARD <qlebastard@gmail.com>
+ * Copyright 2021 Charly Delay <charly@codesink.dev> (@0xcharly)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Publicw License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "charybdis.h"
+
+#ifdef CONSOLE_ENABLE
+#    include "print.h"
+#endif  // CONSOLE_ENABLE
+
+#ifdef POINTING_DEVICE_ENABLE
+#    ifndef CHARYBDIS_MINIMUM_DEFAULT_DPI
+#        define CHARYBDIS_MINIMUM_DEFAULT_DPI 400
+#    endif  // CHARYBDIS_MINIMUM_DEFAULT_DPI
+
+#    ifndef CHARYBDIS_DEFAULT_DPI_CONFIG_STEP
+#        define CHARYBDIS_DEFAULT_DPI_CONFIG_STEP 200
+#    endif  // CHARYBDIS_DEFAULT_DPI_CONFIG_STEP
+
+#    ifndef CHARYBDIS_MINIMUM_SNIPING_DPI
+#        define CHARYBDIS_MINIMUM_SNIPING_DPI 200
+#    endif  // CHARYBDIS_MINIMUM_SNIPER_MODE_DPI
+
+#    ifndef CHARYBDIS_SNIPING_DPI_CONFIG_STEP
+#        define CHARYBDIS_SNIPING_DPI_CONFIG_STEP 100
+#    endif  // CHARYBDIS_SNIPING_DPI_CONFIG_STEP
+
+// Fixed DPI for drag-scroll.
+#    ifndef CHARYBDIS_DRAGSCROLL_DPI
+#        define CHARYBDIS_DRAGSCROLL_DPI 100
+#    endif  // CHARYBDIS_DRAGSCROLL_DPI
+
+#    ifndef CHARYBDIS_DRAGSCROLL_BUFFER_SIZE
+#        define CHARYBDIS_DRAGSCROLL_BUFFER_SIZE 6
+#    endif  // !CHARYBDIS_DRAGSCROLL_BUFFER_SIZE
+
+#    ifndef CHARYBDIS_POINTER_ACCELERATION_FACTOR
+#        define CHARYBDIS_POINTER_ACCELERATION_FACTOR 24
+#    endif  // !CHARYBDIS_POINTER_ACCELERATION_FACTOR
+
+typedef union {
+    uint8_t raw;
+    struct {
+        uint8_t pointer_default_dpi : 4;  // 16 steps available.
+        uint8_t pointer_sniping_dpi : 2;  // 4 steps available.
+        bool    is_dragscroll_enabled : 1;
+        bool    is_sniping_enabled : 1;
+    } __attribute__((packed));
+} charybdis_config_t;
+
+static charybdis_config_t g_charybdis_config = {0};
```

Comment:
This isn't exactly global, and has no way to get these settings outside of this file.  Which can be problematic.  Eg, if you wanted to read the settings to render on an OLED display, for instance.

Reply:
This is `static` on purpose, to avoid these values getting into an inconsistent state/out-of-sync.  You can query these values with the following exposed API:

```c
/** \brief Return the current DPI value for the pointer's default mode. */
uint32_t charybdis_get_pointer_default_dpi(void);

/** \brief Return the current DPI value for the pointer's sniper-mode. */
uint32_t charybdis_get_pointer_sniping_dpi(void);

/** \brief Whether sniper-mode is enabled. */
bool charybdis_get_pointer_sniping_enabled(void);

/** \brief Whether drag-scroll is enabled. */
bool charybdis_get_pointer_dragscroll_enabled(void);
```

Note that `charybdis_get_pointer_default_dpi()` and `charybdis_get_pointer_sniping_dpi()` will return the actual DPI value, instead of the encoded one (which is stored in that structure), which is most likely preferable for users.

Reply:
I somehow missed those 😅 

That said, these should be uint16_t, though.  The max value for it would be 65535, which is well in excess of what can be used.  And using the smaller value saves some room and cycles for AVR chips, which can matter.

Reply:
Makes sense.  I actually didn't realize the "32" in `atmega32u4` was for 32kb of flash, not for 32b.  Should have been obvious in hindsight...